# Visualizing SingleCellExperiment or SpatialExperiment objects


```{r load-lib-2, message=FALSE}
library(BiocParallel)
library(scater)
library(scran)
library(ggplot2)
```

Here we use an example data from a single sample (sample 151673) of human brain dorsolateral prefrontal cortex (DLPFC) in the human brain, measured using the 10x Genomics Visium platform. First, a brief/standard data pre-processing were done with the `scater` and `scran` packages.

```{r setup.preprocess, message=FALSE}
# library(STexampleData)
#
## create ExperimentHub instance
# eh <- ExperimentHub()

## query STexampleData datasets
# myfiles <- query(eh, "STexampleData")
# spe <- myfiles[["EH7538"]]

spe <- readRDS("data/Visium_humanDLPFC.rds")

spe <- addPerCellQC(spe, subsets=list(Mito=grep("^MT-", rowData(spe)$gene_name)))
colData(spe) |> head()

colData(spe) |> data.frame() |> 
  ggplot(aes(x = sum, y = detected, colour = as.factor(in_tissue))) +
   geom_point() 

plotColData(spe, x='sum', y = 'subsets_Mito_percent', other_fields="in_tissue") + facet_wrap(~in_tissue)

```

Firstly, we filter the data to retain the cells that are in the tissue. Then cell-specific biases are normalized using the `computeSumFactors` method.

```{r sce-normalize}

spe <- spe[, spe$in_tissue == 1]

clusters <- quickCluster(
              spe, 
              BPPARAM = BiocParallel::MulticoreParam(workers=2), 
              block.BPPARAM = BiocParallel::MulticoreParam(workers=2)
            )

spe <- computeSumFactors(spe, clusters = clusters, BPPARAM = BiocParallel::MulticoreParam(workers=2))
spe <- logNormCounts(spe)

```

Next, we use the Graph-based clustering method to do the reduction with the `runPCA` and `runTSNE` functions provided in the `scater` package.

```{r sce-reduction}
# identify genes that drive biological heterogeneity in the data set by 
# modelling the per-gene variance
dec <- modelGeneVar(spe)

# Get the top 15% genes.
top.hvgs <- getTopHVGs(dec, prop=0.15)
spe <- runPCA(spe, subset_row=top.hvgs)

output <- getClusteredPCs(reducedDim(spe), BPPARAM = BiocParallel::MulticoreParam(workers=2))
npcs <- metadata(output)$chosen
npcs

reducedDim(spe, "PCAsub") <- reducedDim(spe, "PCA")[,1:npcs,drop=FALSE]

g <- buildSNNGraph(spe, use.dimred="PCAsub", BPPARAM = MulticoreParam(workers=2))
cluster <- igraph::cluster_walktrap(g)$membership
colLabels(spe) <- factor(cluster)
set.seed(123)
spe <- runTSNE(spe, dimred="PCAsub", BPPARAM = MulticoreParam(workers=2))
```



## Dimensional reduction plot

Here, we used the `sc_dim` function provided in the `ggsc` package to visualize the `TSNE` reduction result. Unlike other packages, `ggsc` implemented the `ggplot2` graphic of grammar syntax and visual elements are overlaid through the combinations of graphic layers. The `sc_dim_geom_label` layer is designed to add cell cluster labels to a dimensional reduction plot, and can utilized different implementation of text geoms, such as `geom_shadowtext` in the `shadowtext` package and `geom_text` in the `ggplot2` package (default) through the `geom` argument.


```{r sce-dimplot}
library(ggsc)
library(ggplot2)

sc_dim(spe, reduction = 'TSNE') + sc_dim_geom_label()
sc_dim(spe, reduction = 'TSNE') + 
  sc_dim_geom_label(
    geom = shadowtext::geom_shadowtext,
    color='black', 
    bg.color='white'
  )
```

## Visualize 'features' on a dimensional reduction plot

To visualize the gene expression of cells in the result of reduction, `ggsc` provides `sc_feature` function to highlight on a dimensional reduction plot.


```{r sce-feature}
genes <- c('MOBP', 'PCP4', 'SNAP25', 'HBB', 'IGKC', 'NPY')
target.features <- rownames(spe)[match(genes, rowData(spe)$gene_name)]
sc_feature(spe, target.features[1], slot='logcounts', reduction = 'TSNE')
sc_feature(spe, target.features, slot='logcounts', reduction = 'TSNE')
```

In addition, it provides `sc_dim_geom_feature` layer working with `sc_dim` function to visualize the cells expressed the gene and the cell clusters information simultaneously.

```{r sce-feature-layers}
sc_dim(spe, slot='logcounts', reduction = 'TSNE') +
   sc_dim_geom_feature(spe, target.features[1], color='black')

sc_dim(spe, alpha=.3, slot='logcounts', reduction = 'TSNE') + 
    ggnewscale::new_scale_color() + 
    sc_dim_geom_feature(spe, target.features, mapping=aes(color=features)) +
    scale_color_viridis_d()
```

It also provides `sc_dim_geom_ellipse` to add confidence levels of the the cluster result, and `sc_dim_geom_sub` to select and highlight a specific cluster of cells. 

```{r sce-ellipse}

sc_dim(spe, reduction = 'TSNE') +
  sc_dim_geom_ellipse(level=0.95)

selected.cluster <- c(1, 6, 8)
sc_dim(spe, reduction = 'TSNE') +
  sc_dim_sub(subset=selected.cluster, .column = 'label')

sc_dim(spe, color='grey', reduction = 'TSNE') + 
  sc_dim_geom_sub(subset=selected.cluster, .column = 'label') + 
    sc_dim_geom_label(geom = shadowtext::geom_shadowtext, 
          mapping = aes(subset = label %in% selected.cluster),
            color='black', bg.color='white')  
```


## Dot plot for selected features

```{r sc-dot-spe}
genes <- c('MOBP', 'PCP4', 'SNAP25', 'HBB', 'IGKC', 'NPY')
target.features <- rownames(spe)[match(genes, rowData(spe)$gene_name)]
sc_dot(spe, target.features, slot="logcounts")
```        


## Violin plot of gene expression

`ggsc` provides `sc_violin` to visualize the expression information of specific genes using the violin layer with common legend, the genes can be compared more intuitively.

```{r sce-violin}
sc_violin(spe, target.features[1], slot = 'logcounts')
sc_violin(spe, target.features[1], slot = 'logcounts', 
     .fun=function(d) dplyr::filter(d, value > 0)
     ) + 
     ggforce::geom_sina(size=.1)

sc_violin(spe, target.features, slot = 'logcounts') + 
  theme(axis.text.x = element_text(angle=45, hjust=1))
```

## Spatial features

To visualize the spatial pattern of gene, `ggsc` provides `sc_spatial` to visualize specific features/genes with image information.


```{r sce-spatial, fig.width = 14, fig.height = 10}
library(aplot)
f <- sc_spatial(spe, features = target.features, 
           slot = 'logcounts', ncol = 3, 
           image.mirror.axis = NULL,
           image.rotate.degree = -90
           )

f

pp <- lapply(target.features, function(i) {
  sc_spatial(spe, features = i, slot = 'logcounts', 
          image.rotate.degree = -90, image.mirror.axis = NULL)
})

aplot::plot_list(gglist = pp)
```

```{r sce_spatial-density, fig.width=10, fig.height=10}
f1 <- sc_spatial(spe, features=target.features[1:2], image.rotate.degree=-90) 
f2 <- sc_spatial(spe, features=target.features[1:2], image.rotate.degree=-90, density=TRUE) 

plot_list(f1, f2, ncol=1)
```


```{r sce_spatial-density-join, fig.width=15, fig.height=10}

f3 <- sc_spatial(spe, features=target.features[1:2], image.rotate.degree=-90, 
              density=TRUE, joint=TRUE) 
f4 <- sc_spatial(spe, features=target.features[1:2], image.rotate.degree=-90, 
              density=TRUE, joint=TRUE, common.legend=FALSE)

plot_list(f3, f4, ncol=1)
```



```{r sce-spatial-pie}
sc_spatial(spe, features=target.features, plot.pie=TRUE, 
        pie.radius.scale=.35, linewidth=0.1, image.plot=FALSE)
```
