---
title: "Visualizing single cell data"
author: 
- name: Guangchuang Yu and Shuangbin Xu
  email: guangchuangyu@gmail.com
  affiliation: Department of Bioinformatics, School of Basic Medical Sciences, Southern Medical University
date: "`r Sys.Date()`"
knit: "bookdown::render_book"
site: bookdown::bookdown_site
output:
  bookdown::gitbook:
  bookdown::pdf_book:
    dev: "cairo_pdf"
    latex_engine: xelatex
bibliography: references.bib
biblio-style: apalike
always_allow_html: yes
toc_appendix: yes
fontsize: "12pt"
mainfont: Times New Roman
sansfont: Arial
monofontoptions: "Scale=0.7"
linestretch: 1.5
toc-depth: 2
link-citations: yes
documentclass: book
papersize: A4
classoption: twoside
highlight_bw: yes
geometry: "left=35mm,right=35mm,top=25mm,bottom=25mm"
---


```{r include=FALSE}
library(conflicted)
library(yulab.utils)
library(Seurat)
library(ggplot2)
library(ggsc)

options(width=60)

library(knitr)
opts_chunk$set(message=FALSE, warning=FALSE, eval=TRUE, echo=TRUE, cache=TRUE)
```




# Preface {-}

hello world


<!--chapter:end:index.Rmd-->


# Visualizing Seurat objects

```{r style, echo=FALSE, results="asis", message=FALSE}
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE, 
                      fig.width = 9,
                      fig.height = 6)
library(Seurat)
library(ggplot2)
library(ggsc)                      
```


```{r}
library(Seurat)
dir = "data/filtered_gene_bc_matrices/hg19"

pbmc.data <- Read10X(data.dir = dir)
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells=3, min.features=200)
pbmc

## Normalizing the data
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", 
                      scale.factor = 10000)

pbmc <- NormalizeData(pbmc)

## Identify the 2000 most highly variable genes
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

## In addition we scale the data
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)

pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc), 
               verbose = FALSE)
pbmc <- FindNeighbors(pbmc, dims = 1:10, verbose = FALSE)
pbmc <- FindClusters(pbmc, resolution = 0.5, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:10, umap.method = "uwot", metric = "cosine")

## Run tSNE 
pbmc <- RunTSNE(pbmc, reduction = "pca", dims=1:5, tsne.method = "Rtsne")

## Assigning cell type identity to clusters
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T",
                     "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
```

## Dimensional reduction plot
```{r fig.width=9, fig.height=6}
DimPlot(pbmc, reduction = "umap",
        label = TRUE, pt.size = 0.5) 

library(ggplot2)
library(ggsc)
sc_dim(pbmc) + sc_dim_geom_label()


sc_dim(pbmc) + 
    sc_dim_geom_label(geom = shadowtext::geom_shadowtext, 
            color='black', bg.color='white')
```

## Visualize 'features' on a dimensional reduction plot

```{r}
features = c("MS4A1", "GNLY", "CD3E", 
            "CD14", "FCER1A", "FCGR3A", 
            "LYZ", "PPBP", "CD8A")
FeaturePlot(pbmc,'CD4')
FeaturePlot(pbmc, features)


sc_feature(pbmc, 'CD4')
sc_feature(pbmc, features)
```

Here is the real 'features' on dimensional plot


```{r}
sc_dim(pbmc) + 
   sc_dim_geom_feature(pbmc, 'CD4', color='black')

sc_dim(pbmc, alpha=.3) + 
    ggnewscale::new_scale_color() + 
    sc_dim_geom_feature(pbmc, features, mapping=aes(color=features)) +
    scale_color_viridis_d()

sc_dim(pbmc, reduction = 'tsne') +
    sc_dim_geom_feature(pbmc, 'CD4', color='black')

sc_dim(pbmc, alpha=.3, reduction = 'tsne') +
    ggnewscale::new_scale_color() +
    sc_dim_geom_feature(pbmc, features, mapping=aes(color=features)) +
    scale_color_viridis_d()
```

## Visualize ellipse on a dimensional reduction plot

```{r}
sc_dim(pbmc) +
  sc_dim_geom_ellipse(level=0.95)
```

## Visualize selected clusters 

```{r}
selected <- c("Naive CD4 T", "NK")
sc_dim(pbmc) + 
  sc_dim_sub(subset=selected)
```

```{r}
sc_dim(pbmc, color='grey') + 
  sc_dim_geom_sub(subset=selected) + 
    sc_dim_geom_label(geom = shadowtext::geom_shadowtext, 
          mapping = aes(subset = ident %in% selected),
            color='black', bg.color='white')
```

## Dot plot

```{r eval=FALSE}
DotPlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", 
                           "CD14", "FCER1A", "FCGR3A", 
                           "LYZ", "PPBP", "CD8A"),
        group.by = 'seurat_clusters')
```        


## Violin plot of gene expression

```{r}
VlnPlot(pbmc,'CD4')

sc_violin(pbmc, 'CD4')

## allows applying an user-defined function to transform/filter the data
sc_violin(pbmc, 'CD4', .fun=function(d) dplyr::filter(d, value > 0)) +
  ggforce::geom_sina(size=.1)
```

```{r fig.height=9}
VlnPlot(pbmc, features)
sc_violin(pbmc, features) + 
  theme(axis.text.x = element_text(angle=45, hjust=1))
```



## Spatial features

```{r}
library(SeuratData)
# InstallData("stxBrain")
brain <- LoadData("stxBrain", type = "anterior1")
# Normalization 
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
SpatialFeaturePlot(brain, features = c("Hpca", "Ttr"))


sc_spatial(brain, features = c("Hpca", "Ttr"))
```


<!--chapter:end:single-cell.Rmd-->

