---
title: "Visualizing single cell data"
author: 
- name: Guangchuang Yu and Shuangbin Xu
  email: guangchuangyu@gmail.com
  affiliation: Department of Bioinformatics, School of Basic Medical Sciences, Southern Medical University
date: "`r Sys.Date()`"
knit: "bookdown::render_book"
site: bookdown::bookdown_site
output:
  bookdown::bs4_book:
  bookdown::pdf_book:
    dev: "cairo_pdf"
    latex_engine: xelatex
bibliography: references.bib
biblio-style: apalike
always_allow_html: yes
toc_appendix: yes
fontsize: "12pt"
mainfont: Times New Roman
sansfont: Arial
monofontoptions: "Scale=0.7"
linestretch: 1.5
toc-depth: 2
link-citations: yes
documentclass: book
papersize: A4
classoption: twoside
highlight_bw: yes
geometry: "left=35mm,right=35mm,top=25mm,bottom=25mm"
---


```{r load-libs, include=FALSE}
library(conflicted)
library(yulab.utils)
library(Seurat)
library(ggplot2)
library(ggsc)

options(width=60)

library(knitr)
knitr::opts_chunk$set(tidy = FALSE,
                      warning = FALSE,
                      message = FALSE, 
                      eval=TRUE, 
                      echo=TRUE, 
                      cache=TRUE,
                      fig.width = 9,
                      fig.height = 6,
                      out.width="100%"
    )

```




# Preface {-}

Single-cell RNA sequencing (scRNA-seq) and Spatial RNA sequencing are widely used techniques for profiling gene expression in individual cells with their locations in the histological sections. These allow molecular biology to be studied at a resolution that cannot be matched by bulk sequencing of cell populations. To better visualize the result of reduction, spatial gene expression pattern in single cell or spatial experiment data, `ggsc` provides some layer functions based on the `ggplot2` grammar. It can work with the `SingleCellExperiment` class or `Seurat` class, which are the widely used classes for storing data from single cell experiment.



