<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>2 Visualizing SingleCellExperiment or SpatialExperiment objects | Visualizing single cell data</title>
  <meta name="description" content="2 Visualizing SingleCellExperiment or SpatialExperiment objects | Visualizing single cell data" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="2 Visualizing SingleCellExperiment or SpatialExperiment objects | Visualizing single cell data" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="2 Visualizing SingleCellExperiment or SpatialExperiment objects | Visualizing single cell data" />
  
  
  



<meta name="date" content="2023-11-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="visualizing-seurat-objects.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html"><i class="fa fa-check"></i><b>1</b> Visualizing Seurat objects</a>
<ul>
<li class="chapter" data-level="1.1" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#dimensional-reduction-plot"><i class="fa fa-check"></i><b>1.1</b> Dimensional reduction plot</a></li>
<li class="chapter" data-level="1.2" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#visualize-features-on-a-dimensional-reduction-plot"><i class="fa fa-check"></i><b>1.2</b> Visualize ‘features’ on a dimensional reduction plot</a></li>
<li class="chapter" data-level="1.3" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#visualize-ellipse-on-a-dimensional-reduction-plot"><i class="fa fa-check"></i><b>1.3</b> Visualize ellipse on a dimensional reduction plot</a></li>
<li class="chapter" data-level="1.4" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#visualize-selected-clusters"><i class="fa fa-check"></i><b>1.4</b> Visualize selected clusters</a></li>
<li class="chapter" data-level="1.5" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#dot-plot-for-selected-features"><i class="fa fa-check"></i><b>1.5</b> Dot plot for selected features</a></li>
<li class="chapter" data-level="1.6" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#violin-plot-of-gene-expression"><i class="fa fa-check"></i><b>1.6</b> Violin plot of gene expression</a></li>
<li class="chapter" data-level="1.7" data-path="visualizing-seurat-objects.html"><a href="visualizing-seurat-objects.html#spatial-features"><i class="fa fa-check"></i><b>1.7</b> Spatial features</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><i class="fa fa-check"></i><b>2</b> Visualizing SingleCellExperiment or SpatialExperiment objects</a>
<ul>
<li class="chapter" data-level="2.1" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#dimensional-reduction-plot-1"><i class="fa fa-check"></i><b>2.1</b> Dimensional reduction plot</a></li>
<li class="chapter" data-level="2.2" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#visualize-features-on-a-dimensional-reduction-plot-1"><i class="fa fa-check"></i><b>2.2</b> Visualize ‘features’ on a dimensional reduction plot</a></li>
<li class="chapter" data-level="2.3" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#dot-plot-for-selected-features-1"><i class="fa fa-check"></i><b>2.3</b> Dot plot for selected features</a></li>
<li class="chapter" data-level="2.4" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#violin-plot-of-gene-expression-1"><i class="fa fa-check"></i><b>2.4</b> Violin plot of gene expression</a></li>
<li class="chapter" data-level="2.5" data-path="visualizing-singlecellexperiment-or-spatialexperiment-objects.html"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#spatial-features-1"><i class="fa fa-check"></i><b>2.5</b> Spatial features</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Visualizing single cell data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="visualizing-singlecellexperiment-or-spatialexperiment-objects" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">2</span> Visualizing SingleCellExperiment or SpatialExperiment objects<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#visualizing-singlecellexperiment-or-spatialexperiment-objects" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb29-1" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb29-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb29-2" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb29-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb29-3" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb29-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb29-4" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<p>Here we use an example data from a single sample (sample 151673) of human brain dorsolateral prefrontal cortex (DLPFC) in the human brain, measured using the 10x Genomics Visium platform. First, a brief/standard data pre-processing were done with the <code>scater</code> and <code>scran</code> packages.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-1" tabindex="-1"></a><span class="co"># library(STexampleData)</span></span>
<span id="cb30-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-2" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb30-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-3" tabindex="-1"></a><span class="do">## create ExperimentHub instance</span></span>
<span id="cb30-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-4" tabindex="-1"></a><span class="co"># eh &lt;- ExperimentHub()</span></span>
<span id="cb30-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-5" tabindex="-1"></a></span>
<span id="cb30-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-6" tabindex="-1"></a><span class="do">## query STexampleData datasets</span></span>
<span id="cb30-7"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-7" tabindex="-1"></a><span class="co"># myfiles &lt;- query(eh, &quot;STexampleData&quot;)</span></span>
<span id="cb30-8"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-8" tabindex="-1"></a><span class="co"># spe &lt;- myfiles[[&quot;EH7538&quot;]]</span></span>
<span id="cb30-9"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-9" tabindex="-1"></a></span>
<span id="cb30-10"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-10" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/Visium_humanDLPFC.rds&quot;</span>)</span>
<span id="cb30-11"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-11" tabindex="-1"></a></span>
<span id="cb30-12"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-12" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span><span class="fu">grep</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)))</span>
<span id="cb30-13"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb30-13" tabindex="-1"></a><span class="fu">colData</span>(spe) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 13 columns
##                            barcode_id     sample_id
##                           &lt;character&gt;   &lt;character&gt;
## AAACAACGAATAGTTC-1 AAACAACGAATAGTTC-1 sample_151673
## AAACAAGTATCTCCCA-1 AAACAAGTATCTCCCA-1 sample_151673
## AAACAATCTACTAGCA-1 AAACAATCTACTAGCA-1 sample_151673
## AAACACCAATAACTGC-1 AAACACCAATAACTGC-1 sample_151673
## AAACAGAGCGACTCCT-1 AAACAGAGCGACTCCT-1 sample_151673
## AAACAGCTTTCAGAAG-1 AAACAGCTTTCAGAAG-1 sample_151673
##                    in_tissue array_row array_col
##                    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## AAACAACGAATAGTTC-1         0         0        16
## AAACAAGTATCTCCCA-1         1        50       102
## AAACAATCTACTAGCA-1         1         3        43
## AAACACCAATAACTGC-1         1        59        19
## AAACAGAGCGACTCCT-1         1        14        94
## AAACAGCTTTCAGAAG-1         1        43         9
##                    ground_truth cell_count       sum
##                     &lt;character&gt;  &lt;integer&gt; &lt;numeric&gt;
## AAACAACGAATAGTTC-1           NA         NA       622
## AAACAAGTATCTCCCA-1       Layer3          6      8458
## AAACAATCTACTAGCA-1       Layer1         16      1667
## AAACACCAATAACTGC-1           WM          5      3769
## AAACAGAGCGACTCCT-1       Layer3          2      5433
## AAACAGCTTTCAGAAG-1       Layer5          4      4278
##                     detected subsets_Mito_sum
##                    &lt;numeric&gt;        &lt;numeric&gt;
## AAACAACGAATAGTTC-1       526               37
## AAACAAGTATCTCCCA-1      3586             1407
## AAACAATCTACTAGCA-1      1150              204
## AAACACCAATAACTGC-1      1960              430
## AAACAGAGCGACTCCT-1      2424             1316
## AAACAGCTTTCAGAAG-1      2264              651
##                    subsets_Mito_detected
##                                &lt;numeric&gt;
## AAACAACGAATAGTTC-1                     9
## AAACAAGTATCTCCCA-1                    13
## AAACAATCTACTAGCA-1                    11
## AAACACCAATAACTGC-1                    13
## AAACAGAGCGACTCCT-1                    13
## AAACAGCTTTCAGAAG-1                    12
##                    subsets_Mito_percent     total
##                               &lt;numeric&gt; &lt;numeric&gt;
## AAACAACGAATAGTTC-1              5.94855       622
## AAACAAGTATCTCCCA-1             16.63514      8458
## AAACAATCTACTAGCA-1             12.23755      1667
## AAACACCAATAACTGC-1             11.40886      3769
## AAACAGAGCGACTCCT-1             24.22234      5433
## AAACAGCTTTCAGAAG-1             15.21739      4278</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb32-1" tabindex="-1"></a><span class="fu">colData</span>(spe) <span class="sc">|&gt;</span> <span class="fu">data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb32-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb32-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> sum, <span class="at">y =</span> detected, <span class="at">colour =</span> <span class="fu">as.factor</span>(in_tissue))) <span class="sc">+</span></span>
<span id="cb32-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb32-3" tabindex="-1"></a>   <span class="fu">geom_point</span>() </span></code></pre></div>
<p><img src="ggsc_files/figure-html/setup.preprocess-1.png" width="100%" /></p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb33-1" tabindex="-1"></a><span class="fu">plotColData</span>(spe, <span class="at">x=</span><span class="st">&#39;sum&#39;</span>, <span class="at">y =</span> <span class="st">&#39;subsets_Mito_percent&#39;</span>, <span class="at">other_fields=</span><span class="st">&quot;in_tissue&quot;</span>) <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>in_tissue)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/setup.preprocess-2.png" width="100%" /></p>
<p>Firstly, we filter the data to retain the cells that are in the tissue. Then cell-specific biases are normalized using the <code>computeSumFactors</code> method.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-1" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, spe<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb34-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-2" tabindex="-1"></a></span>
<span id="cb34-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-3" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> <span class="fu">quickCluster</span>(</span>
<span id="cb34-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-4" tabindex="-1"></a>              spe, </span>
<span id="cb34-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-5" tabindex="-1"></a>              <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>), </span>
<span id="cb34-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-6" tabindex="-1"></a>              <span class="at">block.BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>)</span>
<span id="cb34-7"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-7" tabindex="-1"></a>            )</span>
<span id="cb34-8"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-8" tabindex="-1"></a></span>
<span id="cb34-9"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-9" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(spe, <span class="at">clusters =</span> clusters, <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>))</span>
<span id="cb34-10"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb34-10" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span></code></pre></div>
<p>Next, we use the Graph-based clustering method to do the reduction with the <code>runPCA</code> and <code>runTSNE</code> functions provided in the <code>scater</code> package.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-1" tabindex="-1"></a><span class="co"># identify genes that drive biological heterogeneity in the data set by </span></span>
<span id="cb35-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-2" tabindex="-1"></a><span class="co"># modelling the per-gene variance</span></span>
<span id="cb35-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-3" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb35-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-4" tabindex="-1"></a></span>
<span id="cb35-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-5" tabindex="-1"></a><span class="co"># Get the top 15% genes.</span></span>
<span id="cb35-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-6" tabindex="-1"></a>top.hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop=</span><span class="fl">0.15</span>)</span>
<span id="cb35-7"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-7" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row=</span>top.hvgs)</span>
<span id="cb35-8"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-8" tabindex="-1"></a></span>
<span id="cb35-9"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-9" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">getClusteredPCs</span>(<span class="fu">reducedDim</span>(spe), <span class="at">BPPARAM =</span> BiocParallel<span class="sc">::</span><span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>))</span>
<span id="cb35-10"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-10" tabindex="-1"></a>npcs <span class="ot">&lt;-</span> <span class="fu">metadata</span>(output)<span class="sc">$</span>chosen</span>
<span id="cb35-11"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb35-11" tabindex="-1"></a>npcs</span></code></pre></div>
<pre><code>## [1] 13</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-1" tabindex="-1"></a><span class="fu">reducedDim</span>(spe, <span class="st">&quot;PCAsub&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">reducedDim</span>(spe, <span class="st">&quot;PCA&quot;</span>)[,<span class="dv">1</span><span class="sc">:</span>npcs,drop<span class="ot">=</span><span class="cn">FALSE</span>]</span>
<span id="cb37-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-2" tabindex="-1"></a></span>
<span id="cb37-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-3" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">use.dimred=</span><span class="st">&quot;PCAsub&quot;</span>, <span class="at">BPPARAM =</span> <span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>))</span>
<span id="cb37-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-4" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_walktrap</span>(g)<span class="sc">$</span>membership</span>
<span id="cb37-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-5" tabindex="-1"></a><span class="fu">colLabels</span>(spe) <span class="ot">&lt;-</span> <span class="fu">factor</span>(cluster)</span>
<span id="cb37-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-6" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb37-7"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb37-7" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(spe, <span class="at">dimred=</span><span class="st">&quot;PCAsub&quot;</span>, <span class="at">BPPARAM =</span> <span class="fu">MulticoreParam</span>(<span class="at">workers=</span><span class="dv">2</span>))</span></code></pre></div>
<div id="dimensional-reduction-plot-1" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Dimensional reduction plot<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#dimensional-reduction-plot-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we used the <code>sc_dim</code> function provided in the <code>ggsc</code> package to visualize the <code>TSNE</code> reduction result. Unlike other packages, <code>ggsc</code> implemented the <code>ggplot2</code> graphic of grammar syntax and visual elements are overlaid through the combinations of graphic layers. The <code>sc_dim_geom_label</code> layer is designed to add cell cluster labels to a dimensional reduction plot, and can utilized different implementation of text geoms, such as <code>geom_shadowtext</code> in the <code>shadowtext</code> package and <code>geom_text</code> in the <code>ggplot2</code> package (default) through the <code>geom</code> argument.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb38-1" tabindex="-1"></a><span class="fu">library</span>(ggsc)</span>
<span id="cb38-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb38-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb38-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb38-3" tabindex="-1"></a></span>
<span id="cb38-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb38-4" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span> <span class="fu">sc_dim_geom_label</span>()</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-dimplot-1.png" width="100%" /></p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-1" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span> </span>
<span id="cb39-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-2" tabindex="-1"></a>  <span class="fu">sc_dim_geom_label</span>(</span>
<span id="cb39-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-3" tabindex="-1"></a>    <span class="at">geom =</span> shadowtext<span class="sc">::</span>geom_shadowtext,</span>
<span id="cb39-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-4" tabindex="-1"></a>    <span class="at">color=</span><span class="st">&#39;black&#39;</span>, </span>
<span id="cb39-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-5" tabindex="-1"></a>    <span class="at">bg.color=</span><span class="st">&#39;white&#39;</span></span>
<span id="cb39-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb39-6" tabindex="-1"></a>  )</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-dimplot-2.png" width="100%" /></p>
</div>
<div id="visualize-features-on-a-dimensional-reduction-plot-1" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Visualize ‘features’ on a dimensional reduction plot<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#visualize-features-on-a-dimensional-reduction-plot-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To visualize the gene expression of cells in the result of reduction, <code>ggsc</code> provides <code>sc_feature</code> function to highlight on a dimensional reduction plot.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb40-1" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;MOBP&#39;</span>, <span class="st">&#39;PCP4&#39;</span>, <span class="st">&#39;SNAP25&#39;</span>, <span class="st">&#39;HBB&#39;</span>, <span class="st">&#39;IGKC&#39;</span>, <span class="st">&#39;NPY&#39;</span>)</span>
<span id="cb40-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb40-2" tabindex="-1"></a>target.features <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)[<span class="fu">match</span>(genes, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)]</span>
<span id="cb40-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb40-3" tabindex="-1"></a><span class="fu">sc_feature</span>(spe, target.features[<span class="dv">1</span>], <span class="at">slot=</span><span class="st">&#39;logcounts&#39;</span>, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-feature-1.png" width="100%" /></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb41-1" tabindex="-1"></a><span class="fu">sc_feature</span>(spe, target.features, <span class="at">slot=</span><span class="st">&#39;logcounts&#39;</span>, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-feature-2.png" width="100%" /></p>
<p>In addition, it provides <code>sc_dim_geom_feature</code> layer working with <code>sc_dim</code> function to visualize the cells expressed the gene and the cell clusters information simultaneously.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb42-1" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">slot=</span><span class="st">&#39;logcounts&#39;</span>, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span></span>
<span id="cb42-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb42-2" tabindex="-1"></a>   <span class="fu">sc_dim_geom_feature</span>(spe, target.features[<span class="dv">1</span>], <span class="at">color=</span><span class="st">&#39;black&#39;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-feature-layers-1.png" width="100%" /></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb43-1" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">alpha=</span>.<span class="dv">3</span>, <span class="at">slot=</span><span class="st">&#39;logcounts&#39;</span>, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span> </span>
<span id="cb43-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb43-2" tabindex="-1"></a>    ggnewscale<span class="sc">::</span><span class="fu">new_scale_color</span>() <span class="sc">+</span> </span>
<span id="cb43-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb43-3" tabindex="-1"></a>    <span class="fu">sc_dim_geom_feature</span>(spe, target.features, <span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">color=</span>features)) <span class="sc">+</span></span>
<span id="cb43-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb43-4" tabindex="-1"></a>    <span class="fu">scale_color_viridis_d</span>()</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-feature-layers-2.png" width="100%" /></p>
<p>It also provides <code>sc_dim_geom_ellipse</code> to add confidence levels of the the cluster result, and <code>sc_dim_geom_sub</code> to select and highlight a specific cluster of cells.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb44-1" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span></span>
<span id="cb44-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb44-2" tabindex="-1"></a>  <span class="fu">sc_dim_geom_ellipse</span>(<span class="at">level=</span><span class="fl">0.95</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-ellipse-1.png" width="100%" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb45-1" tabindex="-1"></a>selected.cluster <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">6</span>, <span class="dv">8</span>)</span>
<span id="cb45-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb45-2" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span></span>
<span id="cb45-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb45-3" tabindex="-1"></a>  <span class="fu">sc_dim_sub</span>(<span class="at">subset=</span>selected.cluster, <span class="at">.column =</span> <span class="st">&#39;label&#39;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-ellipse-2.png" width="100%" /></p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb46-1" tabindex="-1"></a><span class="fu">sc_dim</span>(spe, <span class="at">color=</span><span class="st">&#39;grey&#39;</span>, <span class="at">reduction =</span> <span class="st">&#39;TSNE&#39;</span>) <span class="sc">+</span> </span>
<span id="cb46-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb46-2" tabindex="-1"></a>  <span class="fu">sc_dim_geom_sub</span>(<span class="at">subset=</span>selected.cluster, <span class="at">.column =</span> <span class="st">&#39;label&#39;</span>) <span class="sc">+</span> </span>
<span id="cb46-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb46-3" tabindex="-1"></a>    <span class="fu">sc_dim_geom_label</span>(<span class="at">geom =</span> shadowtext<span class="sc">::</span>geom_shadowtext, </span>
<span id="cb46-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb46-4" tabindex="-1"></a>          <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">subset =</span> label <span class="sc">%in%</span> selected.cluster),</span>
<span id="cb46-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb46-5" tabindex="-1"></a>            <span class="at">color=</span><span class="st">&#39;black&#39;</span>, <span class="at">bg.color=</span><span class="st">&#39;white&#39;</span>)  </span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-ellipse-3.png" width="100%" /></p>
</div>
<div id="dot-plot-for-selected-features-1" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Dot plot for selected features<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#dot-plot-for-selected-features-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb47-1" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;MOBP&#39;</span>, <span class="st">&#39;PCP4&#39;</span>, <span class="st">&#39;SNAP25&#39;</span>, <span class="st">&#39;HBB&#39;</span>, <span class="st">&#39;IGKC&#39;</span>, <span class="st">&#39;NPY&#39;</span>)</span>
<span id="cb47-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb47-2" tabindex="-1"></a>target.features <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)[<span class="fu">match</span>(genes, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)]</span>
<span id="cb47-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb47-3" tabindex="-1"></a><span class="fu">sc_dot</span>(spe, target.features, <span class="at">slot=</span><span class="st">&quot;logcounts&quot;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sc-dot-spe-1.png" width="100%" /></p>
</div>
<div id="violin-plot-of-gene-expression-1" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> Violin plot of gene expression<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#violin-plot-of-gene-expression-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><code>ggsc</code> provides <code>sc_violin</code> to visualize the expression information of specific genes using the violin layer with common legend, the genes can be compared more intuitively.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb48-1" tabindex="-1"></a><span class="fu">sc_violin</span>(spe, target.features[<span class="dv">1</span>], <span class="at">slot =</span> <span class="st">&#39;logcounts&#39;</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-violin-1.png" width="100%" /></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb49-1" tabindex="-1"></a><span class="fu">sc_violin</span>(spe, target.features[<span class="dv">1</span>], <span class="at">slot =</span> <span class="st">&#39;logcounts&#39;</span>, </span>
<span id="cb49-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb49-2" tabindex="-1"></a>     <span class="at">.fun=</span><span class="cf">function</span>(d) dplyr<span class="sc">::</span><span class="fu">filter</span>(d, value <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb49-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb49-3" tabindex="-1"></a>     ) <span class="sc">+</span> </span>
<span id="cb49-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb49-4" tabindex="-1"></a>     ggforce<span class="sc">::</span><span class="fu">geom_sina</span>(<span class="at">size=</span>.<span class="dv">1</span>)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-violin-2.png" width="100%" /></p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb50-1" tabindex="-1"></a><span class="fu">sc_violin</span>(spe, target.features, <span class="at">slot =</span> <span class="st">&#39;logcounts&#39;</span>) <span class="sc">+</span> </span>
<span id="cb50-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb50-2" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-violin-3.png" width="100%" /></p>
</div>
<div id="spatial-features-1" class="section level2 hasAnchor" number="2.5">
<h2><span class="header-section-number">2.5</span> Spatial features<a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#spatial-features-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To visualize the spatial pattern of gene, <code>ggsc</code> provides <code>sc_spatial</code> to visualize specific features/genes with image information.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-1" tabindex="-1"></a><span class="fu">library</span>(aplot)</span>
<span id="cb51-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-2" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="fu">sc_spatial</span>(spe, <span class="at">features =</span> target.features, </span>
<span id="cb51-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-3" tabindex="-1"></a>           <span class="at">slot =</span> <span class="st">&#39;logcounts&#39;</span>, <span class="at">ncol =</span> <span class="dv">3</span>, </span>
<span id="cb51-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-4" tabindex="-1"></a>           <span class="at">image.mirror.axis =</span> <span class="cn">NULL</span>,</span>
<span id="cb51-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-5" tabindex="-1"></a>           <span class="at">image.rotate.degree =</span> <span class="sc">-</span><span class="dv">90</span></span>
<span id="cb51-6"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-6" tabindex="-1"></a>           )</span>
<span id="cb51-7"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-7" tabindex="-1"></a></span>
<span id="cb51-8"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb51-8" tabindex="-1"></a>f</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-spatial-1.png" width="100%" /></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb52-1" tabindex="-1"></a>pp <span class="ot">&lt;-</span> <span class="fu">lapply</span>(target.features, <span class="cf">function</span>(i) {</span>
<span id="cb52-2"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb52-2" tabindex="-1"></a>  <span class="fu">sc_spatial</span>(spe, <span class="at">features =</span> i, <span class="at">slot =</span> <span class="st">&#39;logcounts&#39;</span>, <span class="at">image.rotate.degree =</span> <span class="sc">-</span><span class="dv">90</span>, <span class="at">image.mirror.axis =</span> <span class="cn">NULL</span>)</span>
<span id="cb52-3"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb52-3" tabindex="-1"></a>})</span>
<span id="cb52-4"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb52-4" tabindex="-1"></a></span>
<span id="cb52-5"><a href="visualizing-singlecellexperiment-or-spatialexperiment-objects.html#cb52-5" tabindex="-1"></a>aplot<span class="sc">::</span><span class="fu">plot_list</span>(<span class="at">gglist =</span> pp)</span></code></pre></div>
<p><img src="ggsc_files/figure-html/sce-spatial-2.png" width="100%" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="visualizing-seurat-objects.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
